### PR TITLE
Changing openmp library from llvm-amdgpu to llvm-openmp 

### PR DESCRIPTION
**Pull request by renjithravindrankannath**
Linking openmp library from rocm-opemnp-extras instead of llvm-amdgpu

### CODE REVIEW

**Comment by renjithravindrankannath** [comment]
Hi @cgmb can you please review this ticket.

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 1]
```diff
@@ -150,6 +150,19 @@ class PyTorch(PythonPackage, CudaPackage):
...
+        depends_on('rocprim@4.5.2')
+        depends_on('hipcub@4.5.2')
+        depends_on('rocthrust@4.5.2')
+        depends_on('roctracer-dev@4.5.2')
+        depends_on('rocrand@4.5.2')
+        depends_on('hipsparse@4.5.2')
+        depends_on('hipfft@4.5.2')
+        depends_on('rocfft@4.5.2')
+        depends_on('hsa-rocr-dev@4.5.2')
+        depends_on('hip@4.5.2')
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/package.py]: These could probably all be `@4.5:`. I don't think they need to be tied to a single release. I've used the Spack py-torch package with ROCm 5.0 and it worked fine.

I'd also be surprised if pytorch depends directly on `hsa-rocr-dev` or `roctracer-dev`, but I could be mistaken.

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 2]
```diff
@@ -303,6 +317,24 @@ def enable_or_disable(variant, keyword='USE', var=None, newer=False):
...
+            env.set('THRUST_PATH', self.spec['rocthrust'].prefix.include)
+            env.set('ROCPRIM_PATH', self.spec['rocprim'].prefix)
+            env.set('HIPRAND_PATH', self.spec['rocrand'].prefix)
+            env.set('ROCRAND_PATH', self.spec['rocrand'].prefix)
+            env.set('MIOPEN_PATH', self.spec['miopen-hip'].prefix)
+            env.set('RCCL_PATH', self.spec['rccl'].prefix)
+            env.set('HIPCUB_PATH', self.spec['hipcub'].prefix)
+            env.set('ROCTHRUST_PATH', self.spec['rocthrust'].prefix)
+            env.set('ROCTRACER_PATH', self.spec['roctracer-dev'].prefix)
+            env.set('RCCL_INCLUDE_DIR', self.spec['rccl'].prefix.include)
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/package.py]: Ugh. Spack adds all these libraries to the `CMAKE_PREFIX_PATH` automatically. Using a library like rocblas _should_ be as easy as `find_package(rocblas REQUIRED)` and then `target_link_libraries(<whatever> PRIVATE roc::rocblas)` in pytorch.

It's unfortunate how weird this code is, but I suppose this makes sense given [how pytorch is looking for the libraries](https://github.com/pytorch/pytorch/blob/f83d047338f98b2aa2201c8d26d7baba563c38f6/cmake/public/LoadHIP.cmake). It seems that complexity begets complexity.

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+@@ -1280,6 +1280,12 @@ if(USE_ROCM)
+     /opt/rocm/hcc/include
+     /opt/rocm/rocblas/include
+     /opt/rocm/hipsparse/include
++    ${ROCPRIM_PATH}/include
++    ${ROCRAND_PATH}/rocrand/include
++    ${HIPCUB_PATH}/include
++    ${HIPFFT_PATH}/hipfft/include
++    ${RCCL_INCLUDE_DIR}
++    ${ROCRAND_PATH}
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: The existing code here that hardcodes the `/opt/rocm` include directories will cause nothing but headaches. The code you're adding is better, but you're probably going to have some troubles a few releases down the line if you specify the include directories like that, too. The paths you're specifying may change, so it's better to use the variables from CMake.

`find_package` creates standard variables containing the include directories. If you do a search through `<rocmdir>/lib/cmake/` for `INCLUDE_DIRS` and you'll find the names of the variables to use. e.g.
```
${rocblas_INCLUDE_DIRS}
${hipsparse_INCLUDE_DIRS}
${rocprim_INCLUDE_DIRS}
${rocrand_INCLUDE_DIRS}
${hipcub_INCLUDE_DIRS}
${hipfft_INCLUDE_DIRS}
${rccl_INCLUDE_DIRS}
```

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+   set_target_properties(caffe2_pybind11_state PROPERTIES LINK_FLAGS "${_caffe2_pybind11_state_linker_flags}")
+   target_include_directories(caffe2_pybind11_state PRIVATE $<INSTALL_INTERFACE:include>)
+   target_include_directories(caffe2_pybind11_state PRIVATE ${Caffe2_CPU_INCLUDE})
++  target_include_directories(caffe2_pybind11_state PRIVATE
++      ${ROCPRIM_PATH}/include
++      ${ROCRAND_PATH}/rocrand/include
++      ${HIPCUB_PATH}/include
++      ${HIPFFT_PATH}/hipfft/include
++      ${RCCL_INCLUDE_DIR}
++      ${ROCRAND_PATH})
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: Same comment as on `torch_hip`.

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+       target_link_libraries(caffe2_pybind11_state_hip ${PYTHON_LIBRARIES})
+diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
+index cf37984c32..65a86b1c4f 100644
+--- a/cmake/Dependencies.cmake
++++ b/cmake/Dependencies.cmake
+@@ -1321,6 +1321,7 @@ if(USE_ROCM)
+   include_directories(SYSTEM ${HIPRAND_PATH}/include)
+   include_directories(SYSTEM ${ROCRAND_PATH}/include)
+   include_directories(SYSTEM ${THRUST_PATH})
++  include_directories(SYSTEM ${RCCL_INCLUDE_DIR})
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: Same comment as on `torch_hip`.

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+@@ -31,6 +31,13 @@ if(BUILD_CAFFE2_OPS)
+     torch_set_target_props(caffe2_detectron_ops_hip)
+     target_compile_options(caffe2_detectron_ops_hip PRIVATE ${HIP_CXX_FLAGS})
+     target_link_libraries(caffe2_detectron_ops_hip torch)
++    target_include_directories(caffe2_detectron_ops_hip PRIVATE
++                              ${ROCPRIM_PATH}/include
++                              ${ROCRAND_PATH}/rocrand/include
++                              ${HIPCUB_PATH}/include
++                              ${HIPFFT_PATH}/hipfft/include
++                              ${ROCRAND_PATH})
```
Comment (cgmb) [inline var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: Same comment as on `torch_hip`.

**Review by cgmb — COMMENTED** [review]
@haampie and @srekolam are really the ones that should be looking at this, but I left some comments.

One thing that stands out to me is that you seem to have two sets of changes here. One change is to make Spack's py-torch recipe depend on the Spack ROCm packages. IMO, that should probably be split off into its own PR, because it will have a major impact on users. It's a good thing to do, but it completely changes the setup instructions for installing a ROCM-enabled version of py-torch through Spack.

The other change is what corresponds to the title: using rocm-openmp-extras instead of the openmp library in llvm-amdgpu. I know almost nothing about OpenMP, so I have very little to say about that.

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch (hunk 6 above)
Comment (renjithravindrankannath) [reply var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: @cgmb, 
py-torch changes were not intented to include in this PR. It got added by mistake when I did fetch and merge recently from my local repo. Sorry for that. I will split py-torch changes and keep only the changes required only for openmp library shortly.

File: var/spack/repos/builtin/packages/py-torch/package.py (hunk 1 above)
Comment (renjithravindrankannath) [reply var/spack/repos/builtin/packages/py-torch/package.py]: Hi @cgmb ,
Could you please share the build command that used when the Spack py-torch packaging worked fine with ROCm 5.0.

File: var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch (hunk 6 above)
Comment (renjithravindrankannath) [reply var/spack/repos/builtin/packages/py-torch/py-torch-rocm-003.patch]: Hi @cgmb ,
I removed unwanted changes for for openmp library. Let us know if it is good to approve.

**Comment by cgmb** [comment]
I suppose my question is: why are changes needed to Tensile? I talked with some Tensile devs and they have no idea what rocm-openmp-extras is.

As far as I understand, the purpose of the special ROCm OpenMP stack is to enable OpenMP GPU offloading, but I don't think that's needed or desired for Tensile. Or, is this required even just to make basic CPU OpenMP work with llvm-amdgpu?

**Comment by renjithravindrankannath** [comment]
This will be tracked under new PR #30883

**Pull request closed without merging**